sled-agent/src/rack_setup/plan/service.rs

```diff
@@ -0,0 +1,354 @@
+
+//! Plan generation for "where should services be initialized".
+
+use crate::params::{
+    DatasetEnsureBody, ServiceType, ServiceZoneRequest, ZoneType,
+};
+use crate::rack_setup::config::SetupServiceConfig as Config;
+use omicron_common::address::{
+    get_switch_zone_address, Ipv6Subnet, ReservedRackSubnet, DNS_PORT,
+    DNS_SERVER_PORT, RSS_RESERVED_ADDRESSES, SLED_PREFIX,
+};
+use omicron_common::backoff::{
+    internal_service_policy, retry_notify, BackoffError,
+};
+use serde::{Deserialize, Serialize};
+use sled_agent_client::{
+    types as SledAgentTypes, Client as SledAgentClient, Error as SledAgentError,
+};
+use slog::Logger;
+use std::collections::HashMap;
+use std::net::{Ipv6Addr, SocketAddrV6};
+use std::path::{Path, PathBuf};
+use thiserror::Error;
+use uuid::Uuid;
+
+// The number of Nexus instances to create from RSS.
+const NEXUS_COUNT: usize = 1;
+
+// The number of CRDB instances to create from RSS.
+const CRDB_COUNT: usize = 1;
+
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove
+// when Nexus provisions Oximeter.
+const OXIMETER_COUNT: usize = 1;
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove
+// when Nexus provisions Clickhouse.
+const CLICKHOUSE_COUNT: usize = 1;
+// TODO(https://github.com/oxidecomputer/omicron/issues/732): Remove.
+// when Nexus provisions Crucible.
+const MINIMUM_ZPOOL_COUNT: usize = 3;
+
+fn rss_service_plan_path() -> PathBuf {
+    Path::new(omicron_common::OMICRON_CONFIG_PATH).join("rss-service-plan.toml")
+}
+
+/// Describes errors which may occur while generating a plan for services.
+#[derive(Error, Debug)]
+pub enum PlanError {
+    #[error("I/O error while {message}: {err}")]
+    Io {
+        message: String,
+        #[source]
+        err: std::io::Error,
+    },
+
+    #[error("Cannot deserialize TOML file at {path}: {err}")]
+    Toml { path: PathBuf, err: toml::de::Error },
+
+    #[error("Error making HTTP request to Sled Agent: {0}")]
+    SledApi(#[from] SledAgentError<SledAgentTypes::Error>),
+
+    #[error("Error initializing sled via sled-agent: {0}")]
+    SledInitialization(String),
+
+    #[error("Failed to construct an HTTP client: {0}")]
+    HttpClient(reqwest::Error),
+}
+
+#[derive(Clone, Debug, Default, Deserialize, Serialize, PartialEq)]
+pub struct SledRequest {
+    /// Datasets to be created.
+    #[serde(default, rename = "dataset")]
+    pub datasets: Vec<DatasetEnsureBody>,
+
+    /// Services to be instantiated.
+    #[serde(default, rename = "service")]
+    pub services: Vec<ServiceZoneRequest>,
+
+    /// DNS Services to be instantiated.
+    #[serde(default, rename = "dns_service")]
+    pub dns_services: Vec<ServiceZoneRequest>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Plan {
+    pub services: HashMap<SocketAddrV6, SledRequest>,
+}
+
+impl Plan {
+    pub async fn load(log: &Logger) -> Result<Option<Plan>, PlanError> {
+        // If we already created a plan for this RSS to allocate
```

Comment:
I just wan't to note that eventually, this should probably be written and retreived from the bootstore for fault tolerance. The Handoff "switch" can go there also. Obviously there's a lot more work to do on the bootstore front, and I'm not looking to change anything here now (certainly not before MUP) but I wanted to leave a note.

Reply:
I might need a better understanding of the bootstore - how is it expected to be replicated within the rack?

Reply:
Unfortunately bootstore is not well documented yet. It's a replacement for the more complicated reconfiguration protocol in [RFD 238 section 5](https://rfd.shared.oxide.computer/rfd/0238#_trust_quorum_reconfiguration) and is alluded to in RFD 301 with some implied functionality. Once we get past MUP I'll write it up completely in section 5 of RFD 238 or a new RFD.

For now though, you can think of it the following way. There's a coordinator (based in the boot agent) that drives a 2-phase commit with input from Nexus or RSS, given initial setup or standard reconfiguration. Replication is to every sled with the exception of specific key-shares going only to their assigned replicas. Transport is over sprockets.

The intended uses are:
 * Storage of key shares and trust quorum metadata
 * Storage of enough networking information required to enable NTP and bring up CockroachDB
 * Storage of initial plan data from RSS that can be fed into Nexus/CockroachDB on the nexus handoff. This should allow the system to tolerate failures of this initial setup in a more robust manner and prevent conflicting plans in case of those versions. John and I discussed this a bit a while back, but right now it's just a rough idea in my head. Since the plan and key-shares are intimately tied together for rack init, I figured we'd just ship them together from RSS when the user pressed go.

Currently the protocol for the bootstore implementation has been started with a focus on storing key shares. There are some comments related to also storing the plan, but nothing implemented yet. Those changes should be relatively minor from the bootstore perspective.

Reply:
Thanks for that summary @andrewjstone!